### ISSUE TITLE
Json.net fails on constructor in Rec2f struct

### ISSUE THREAD

**Issue by Shadowblitz16** [JamesNK/Newtonsoft.Json#2527]
### Source/destination types

https://github.com/Shadowblitz16/MicroEngine
`MicroEngine/Defines/Structs.cs`

### Source/destination JSON
```
{
  "Image": "Assets/Sprite.png",
  "Index": 1,
  "Speed": 60,
  "Frames": [
    {
      "Pos": {
        "X": 0,
        "Y": 0
      },
      "Size": {
        "X": 0,
        "Y": 0
      }
    },
    {
      "Pos": {
        "X": 0,
        "Y": 0
      },
      "Size": {
        "X": 0,
        "Y": 0
      }
    }
  ],
  "Origin": {
    "X": 0.0,
    "Y": 0.0
  },
  "Blend": {
    "R": 255,
    "G": 255,
    "B": 255,
    "A": 255
  },
  "Position": {
    "X": 256.0,
    "Y": 256.0
  },
  "Rotation": 18.10001,
  "Scale": {
    "X": 4.0,
    "Y": 4.0
  }
}
```

### Expected behavior

I would think that it could serialize Sprite's `List<Rec2i> Frames { get; }`

### Actual behavior

It populates but it never actually sets the Vec2i's `Pos` and `Size` they are always 0,0

### Steps to reproduce

Asset.cs holds a Load and Save method which I use json.net in
the best way to produce is to..
- Copy the json to `MicroEngine/MicroEngineTest/bin/Debug/net5.0/Assets/Sprite.json`
- Load up the Project
- Make sure your calling Load instead of the parameter list in WindowTest.cs
- Fire it up and pause when its in the main loop
- Open up Sprite in the inspector and see that everything else works but the List was populated with empty Vec2i's

**Comment by elgonzo** [comment]
Yeah, from the look of it, it seems to be a bug / design mistake in `DefaultContractResolver.GetImmutableConstructor`:

https://github.com/JamesNK/Newtonsoft.Json/blob/e44ab3348fabcbe9ed6b5053b706a9e10b945ba1/Src/Newtonsoft.Json/Serialization/DefaultContractResolver.cs#L629-L656

Instead of iterating over all available parameterized constructors of immutable types such as your struct to find one whose arguments match the properties, it just picks the first one ("first" as in the first provided through reflection). And if that doesn't match the properties it's game over, despite there being other public constructors whose arguments would match the properties. 

And in case of `Rec2i` (and other structs), the first parameterized constructor provided through reflection is the "wrong" `public Rec2i(int x, int y, int w, int h)` constructor. Bumma!

I do NOT suggest to rely on the order of constructors in your source code. There is no guarantee whatsoever that the order of the constructors provided through reflection will be the same as the order they appear in the source code. I don't even know if the order in which constructors are provided through reflection will always be the same, of if this order arbitrary and undetermined.
<br>


There are basically two ways out of this hole: Either annotate the appropiate constructors in your immutable structs with `[JsonConstructor]`  attributes, kind like:

```c#
public struct Rec2i
{
    public Vec2i Pos  { get;}
    public Vec2i Size { get;}

    public Rec2i(int x, int y, int w, int h)
    {
        Pos  = new Vec2i(x, y);
        Size = new Vec2i(w, h);
    }

    [JsonConstructor]
    public Rec2i(Vec2i pos, Vec2i size)
    {
        Pos  = pos;
        Size = size;
    }

       ...
}
```

Or, if annotating your data models with Newtonsoft.Json-specific attributes is not desirable, derive your own ContractResolver from DefaultContractResolver that selects the correct constructor for your immutable structs. For getting a basic idea of how to do this, look at some answers of this Stackoverflow question: https://stackoverflow.com/questions/23017716/json-net-how-to-deserialize-without-using-the-default-constructor But note that the contract resolvers given in those answers are not necessarily ready-to-use -- you probably need to adapt them so that the constructors that are correct for your use case are selected.
